apis/python/src/tiledbsoma/tiledb_object.py

```diff
@@ -113,15 +125,100 @@ def exists(self) -> bool:
+        opened, the object should later be closed to release such resources. This can be automated
+        with a context manager, e.g.:
+
+          with tiledbsoma.SparseNDArray(uri=...).open() as X:
+              data = X.read_table(...)
+              data = X.read_table(...)
+
+        If the object is not opened, then any individual read or write operation effectively opens
+        and closes it transiently for the operation.
+        """
+        if mode not in ("r", "w"):
+            raise ValueError(
+                self.__class__.__name__ + " open mode must be one of 'r', 'w'"
+            )
+        if self._open_mode:
+            raise RuntimeError(self.__class__.__name__ + " is already open")
+        self._open_mode = mode
+        return self
+
+    def close(self) -> None:
+        """
+        Release any resources held while the object is open. Closing an already-closed object is a
+        no-op.
+        """
+        if self._open_mode:
+            self._open_mode = ""
+            self._close_stack.close()
+
+    def __enter__(self: TTileDBObject) -> TTileDBObject:
+        assert (
+            self._open_mode
+        ), f"use {self.__class__.__name__}.open() as context manager"
+        return self
+
+    def __exit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc_value: Optional[BaseException],
+        traceback: Optional[TracebackType],
+    ) -> None:
+        if self._open_mode:
+            self.close()
+
+    @contextmanager
+    def _maybe_open(self, mode: str = "r") -> Iterator[None]:
+        """
+        Internal helper context for read/write methods to open self just for the duration of the
+        operation -- IFF self isn't already open. Also rejects attempts to write when self is
+        already open read-only.
+
+        when mode == "w":
+            if self is already open in read-only mode, raise an error.
+            if self is already open in w mode, do nothing.
+            otherwise open("w") until context exit.
+        when mode == "r":
+            if self is already open (IN EITHER MODE*), do nothing.
+            otherwise open("r") until context exit.
+
+            * We don't necessarily reject reads while we're open for writing, and different
+              timestamps may be used in this case. Usually, reads while open for writing just
+              pertain to metadata (e.g. exists()). The underlying TileDB object may reject other
+              cases.
+        """
+        assert mode in ("r", "w")
+        if self._open_mode:
+            if self._open_mode == "r" and mode == "w":
+                raise RuntimeError(
+                    self.__class__.__name__ + " is already open read-only"
+                )
+            yield
+        else:
+            with self.open(mode):
+                yield
+
+    def __del__(self) -> None:
+        self.close()
```

Comment:
Need to call __del__ on super() if it exists, as this definitely may get specialized in a sub-class.

Reply:
There's no default `__del__` to call unless we know the superclass definitely provides one; left a comment to this effect. `__del__` is weird in general with a [tortured history](https://peps.python.org/pep-0442/), I'm not sure if it's conventional is to try to compensate for developer forgetting to close like this.

Reply:
I am going to defer to @thetorpedodog on this one.  In other places in our code base, we do some fairly convoluted stuff to handle this edge case, eg, 

```
    def __del__(self) -> None:
        """Ensure that we're closed when our last ref disappears."""
        # If any superclass in our MRO has a __del__, call it.
        sdel = getattr(super(), "__del__", lambda: None)
        sdel()
```